src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt

```diff
@@ -0,0 +1,95 @@
+        override fun visitUseSpeck(o: RsUseSpeck) {
+            if (!holder.project.isNewResolveEnabled) return
+            val item = o.ancestorStrict<RsUseItem>() ?: return
+            if (item.isReexport) return
+            if (o.path?.resolveStatus != PathResolveStatus.RESOLVED) return
+
+            // Do not check uses if there is a child mod inside the current mod
+            val parentMod = o.containingMod
+            if (parentMod.expandedItemsExceptImplsAndUses.any { it is RsModItem || it is RsModDeclItem }) return
+
+            checkUseSpeck(o, holder)
+        }
+    }
+
+    private fun checkUseSpeck(useSpeck: RsUseSpeck, holder: RsProblemsHolder) {
+        if (!useSpeck.isUsed()) {
+            val element = getHighlightElement(useSpeck)
+            holder.registerLintProblem(
+                element,
+                "Unused import: `${useSpeck.text}`"
+            )
+        }
+    }
+}
+
+private fun getHighlightElement(useSpeck: RsUseSpeck): PsiElement {
+    if (useSpeck.parent is RsUseItem) {
+        return useSpeck.parent
+    }
+    return useSpeck
+}
+
+/**
+ * Returns true if this use speck has at least one usage in its containing owner (module or function).
+ * A usage can be either a path that uses the import of the use speck or a method call/associated item available through
+ * a trait that is imported by this use speck.
+ */
+private fun RsUseSpeck.isUsed(): Boolean {
+    val owner = this.parentOfType<RsUseItem>()?.parent as? RsItemsOwner ?: return true
+    val usage = owner.pathUsage
+    return isUseSpeckUsed(this, usage)
+}
+
+private fun isUseSpeckUsed(useSpeck: RsUseSpeck, usage: PathUsageMap): Boolean {
+    // Use speck with an empty group is always unused
+    val useGroup = useSpeck.useGroup
+    if (useGroup != null && useGroup.useSpeckList.isEmpty()) return false
+
+    val (candidateItems: Set<RsElement>, usedItems: Set<RsElement>) = if (useSpeck.isStarImport) {
+        val module = useSpeck.path?.reference?.resolve() as? RsMod ?: return true
+
+        // Extract all visible items from the target module
+        val candidateItems = module.exportedItems(useSpeck.containingMod)
+
+        // Find path usages with names of the extracted items
+        val usedItems = candidateItems.mapNotNull {
+            usage.pathUsages[it.name]
+        }.fold(mutableSetOf<RsElement>()) { acc, value ->
+            acc.addAll(value)
+            acc
+        }
+
+        Pair(candidateItems.map { it.item }.toSet(), usedItems)
+    } else {
+        val path = useSpeck.path ?: return true
+        val items = path.reference?.multiResolve() ?: return true
+
+        val name = useSpeck.itemName(withAlias = true) ?: return true
+        val usedItems = usage.pathUsages[name].orEmpty()
+
+        Pair(items.toSet(), usedItems)
+    }
+
+    return candidateItems.intersects(usedItems) || candidateItems.intersects(usage.traitUsages)
```

Comment:
It was a bit hard for me to understand the logic with intersection

Also I am not sure that we should unite `pathUsages[it.name]`, because it may lead to false-negatives (please add such test):

```rust
mod mod1 {
    pub fn foo() {}
}
mod mod2 {
    pub use crate::mod1::foo;
    pub fn bar() {}
}
mod test {
    use crate::mod2::*;
    use crate::mod1::foo as bar;
    fn main() {
        bar();
    }
}
```

I would do something like this:
```kotlin
    val items = if (useSpeck.isStarImport) {
        val module = useSpeck.path?.reference?.resolve() as? RsMod ?: return true
        module.exportedItems(useSpeck.containingMod)
    } else {
        val path = useSpeck.path ?: return true
        val items = path.reference?.multiResolve() ?: return true
        val name = useSpeck.itemName(withAlias = true) ?: return true
        items.filterIsInstance<RsNamedElement>().map { NamedItem(name, it) }
    }

    return items.any { (name, item) ->
        item in usage.pathUsages[name].orEmpty()
            || item in usage.traitUsages
    }
```

Reply:
Thanks!